### PR TITLE
 Improve glslES3.00  shader syntax compatibility

### LINKED ISSUE #2047: glsl compile error with GLSL ES 3.00
**Describe the bug**
I think it's a glsl code compile bug. Support for GLSL ES 3.00 is not very good.

**To Reproduce**

The source code of frag shader:
```glsl
precision highp float;

in vec4 vColor;
in vec2 vPosition;
out vec4 fragColor;

void main () {
  float A = -dot(vPosition, vPosition);
  if (A < -4.0) discard;
  float B = exp(A) * vColor.a;
  fragColor = vec4(B * vColor.rgb, B);
}
```

The [ShaderFactory.convertTo300](https://github.com/galacean/engine/blob/6c0639468a63c3d83116f03088e9c01a27f54014/packages/core/src/shaderlib/ShaderFactory.ts#L64) method converts the code to:

```glsl
  #version 300 es
  #define GRAPHICS_API_WEBGL2
....
 
 precision highp float;

 in vec4 vColor;
 in vec2 vPosition;
 out vec4 fragColor;
 
 // This line is NOT I expect:
 out vec4 glFragColor;

 void main() {
   float A = -dot(vPosition, vPosition);
   if (A < -4.0) discard;
   float B = exp(A) * vColor.a;
   fragColor = vec4(B * vColor.rgb, B);
 }

``` 

**Screenshots**

image

**Expected behavior**

1. Keep the GLSL ES 3.00 code unchanged. OR
2. Covert to a lower version can compile.

Perhaps giving the user the option to choose would be a wise decision.

> PS: String replacement is an unsafe approach for handling GLSL syntax. But I have no idea if there's a better approach to solve the problem.

### PR DESCRIPTION
**Pull request by zhuxudong**
### Please check if the PR fulfills these requirements
Fix #2047 
- [x] The commit message follows our [guidelines](https://github.com/galacean/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### Other information:

### CODE REVIEW

**Review by eyworldwide — APPROVED** [review]

File: packages/core/src/shader/ShaderPass.ts [hunk 1]
```diff
@@ -130,7 +130,7 @@ export class ShaderPass extends ShaderPart {
         ${macroNameStr}
       ` + ShaderFactory.parseIncludes(this._fragmentSource);
 
-    if (isWebGL2) {
+    if (isWebGL2 && !ShaderFactory._isGLSL300(fragmentSource)) {
```
Comment (GuoLei1990) [inline packages/core/src/shader/ShaderPass.ts:133]: Atomization is better，like three.js

**Review by GuoLei1990 — CHANGES_REQUESTED** [review]
Add unit test

**Comment by GuoLei1990** [comment]
Completed supplementary inspection：
image

**Comment by GuoLei1990** [comment]
image
Babylon.js don't have this bug

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 2]
```diff
@@ -47,28 +49,42 @@ export class ShaderFactory {
...
-        shader = shader.replace(/\bgl_FragColor\b/g, "gl_FragData[0]");
-        const result = shader.match(/\bgl_FragData\[.+?\]/g);
-        shader = this._replaceMRTShader(shader, result);
-      } else {
-        shader = shader.replace(/void\s+?main\s*\(/g, `out vec4 glFragColor;\nvoid main(`);
-        shader = shader.replace(/\bgl_FragColor\b/g, "glFragColor");
+      shader = shader.replace(/\btexture(2D|Cube)LodEXT\b/g, "textureLod");
+      shader = shader.replace(/\btexture(2D|Cube)GradEXT\b/g, "textureGrad");
+      shader = shader.replace(/\btexture2DProjLodEXT\b/g, "textureProjLod");
+      shader = shader.replace(/\btexture2DProj\b/g, "textureProj");
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:60]: Put this line firast

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 3]
```diff
@@ -47,28 +49,42 @@ export class ShaderFactory {
    * @param isFrag - Whether it is a fragment shader.
    * */
   static convertTo300(shader: string, isFrag?: boolean) {
-    /** replace attribute and in */
     shader = shader.replace(/\battribute\b/g, "in");
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:52]: Only vertex need

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 4]
```diff
@@ -47,28 +54,42 @@ export class ShaderFactory {
...
     shader = shader.replace(/\bvarying\b/g, isFrag ? "in" : "out");
-
-    /** replace api */
     shader = shader.replace(/\btexture(2D|Cube)\b/g, "texture");
-    shader = shader.replace(/\btexture(2D|Cube)LodEXT\b/g, "textureLod");
+    shader = shader.replace(/\btexture2DProj\b/g, "textureProj");
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:60]: Test

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 5]
```diff
@@ -3,10 +3,17 @@ import { ShaderLib } from "./ShaderLib";
 
 export class ShaderFactory {
   /** @internal */
-  static readonly _shaderExtension = ["GL_EXT_shader_texture_lod", "GL_OES_standard_derivatives", "GL_EXT_draw_buffers"]
+  static readonly _shaderExtension = [
+    "GL_EXT_shader_texture_lod",
+    "GL_OES_standard_derivatives",
+    "GL_EXT_draw_buffers",
+    "GL_EXT_frag_depth"
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:10]: WebGL don't need this

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 6]
```diff
@@ -47,28 +54,42 @@ export class ShaderFactory {
...
     }
 
     return shader;
   }
 
+  /**
+   * The temporary solution for checking whether the fragment shader is GLSL 300 es.
+   * @internal
+   */
+  static _hasOutput(fragmentShader: string) {
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:89]: - private
- `_hasES3Output`

File: packages/core/src/shaderlib/ShaderFactory.ts [hunk 7]
```diff
@@ -3,10 +3,17 @@ import { ShaderLib } from "./ShaderLib";
...
+    "GL_EXT_shader_texture_lod",
+    "GL_OES_standard_derivatives",
+    "GL_EXT_draw_buffers",
+    "GL_EXT_frag_depth"
+  ]
     .map((e) => `#extension ${e} : enable\n`)
     .join("");
 
+  private static readonly _hasOutInFragReg = /\bout\s+(?:\w+\s+)?(?:float|vec4)\s+(?:\w+)\s*;/; // [layout(location = 0)] out [highp] [vec4|float] [color|gl_FragDepth];
+
```
Comment (GuoLei1990) [inline packages/core/src/shaderlib/ShaderFactory.ts:16]: Opt comments

**Review by GuoLei1990 — APPROVED** [review]

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**